README.md
# discon


[![Build Status](https://travis-ci.org/trinker/discon.png?branch=master)](https://travis-ci.org/trinker/discon)
[![Coverage Status](https://coveralls.io/repos/trinker/discon/badge.png)](https://coveralls.io/r/trinker/discon)
[![DOI](https://zenodo.org/badge/5398/trinker/discon.svg)](http://dx.doi.org/10.5281/zenodo.15711)
<a href="https://img.shields.io/badge/Version-0.2.0-orange.svg"><img src="https://img.shields.io/badge/Version-0.2.0-orange.svg" alt="Version"/></a></p>


<img src="inst/discon_logo/r_discon.png" width = "300" alt="discon logo">   


[discon](http://trinker.github.com/discon_dev) is a collection of R tools to enhance analysis of  *discourse connectors* in text.  Discourse connectors are cohesive devices that can be used to help identify themes within a text.  This package provides computational means of extracting and visualizing various elements from the text that contain discourse connectors.  This can assist in qualitative analysis of discourse by identifying categories that may aide analysis (using the computer for efficiency and data coverage) towards generating themes.

> Discourse connectors are devices used to bridge between turns (in speech) and sentences, indicating the logical relations among the parts of a the logical relations among the parts of a framework for the listener/reader.  There are two major classes of discourse connectors: discourse markers and linking adverbials. Discourse markers &ndash; forms like *ok*, *well*, and *now* &ndash; are restricted primarily to spoken discourse. These forms have distinct discourse functions, but it is difficult to identify the specific meaning of the word itself. In contrast, linking adverbials &ndash; forms like *however*, *thus*, *therefore*, *for example* (*e.g.*), and *that is* (*i.e.*) &ndash; are found in both spoken and written registers, and they have greater inherent meaning than discourse markers. (Biber, 2006, p. 66)

Please see the following resources for additional information:

- Alemany, L. A. (2005). [Representing discourse for automatic text summarization via shallow NLP techniques](http://www.cs.famaf.unc.edu.ar/~laura/shallowdisc4summ/tesi_electronica.pdf) (Unpublished doctoral dissertation). Universitat de Barcelona, Barcelona.     
  - [Web Site](http://russell.famaf.unc.edu.ar/~laura/shallowdisc4summ/discmar)    
- Biber, D. (2006). *University language: A corpus-based study of spoken and written registers*. Amsterdam: John Benjamins Publishing Company.
- Halliday, M. A. K. & Hasan, R. (1976). *Cohesion in English*. English Language Series. London: Longman.
- Kalajahi, S. A. R., Abdullah, A. N., Mukundan, J., & Tannacito, D. J. (2012). [Discourse connectors: An overview of the history, definition and classification of the term](http://goo.gl/eS0OwV). *World Applied Sciences Journal, 19*(11), 1659-1673. 
- Linguistic Data Consortium. (2004) [*Simple metadata annotation specification (MDE)*](http://www.itl.nist.gov/iad/mig/tests/rt/2004-fall/docs/SimpleMDE_V6.2-draft.pdf). University of Pennsylvania.  Retrieved from [http://www.itl.nist.gov/iad/mig/tests/rt/2004-fall/docs/SimpleMDE_V6.2-draft.pdf](http://www.itl.nist.gov/iad/mig/tests/rt/2004-fall/docs/SimpleMDE_V6.2-draft.pdf)
- Ryan, G. W. & Bernard, H. R. (2003). [Techniques to identify themes](http://goo.gl/KdYxB8). *Field Methods.  15*(1), 85-109. doi:[10.1177/1525822X02239569.](http://fmx.sagepub.com/content/15/1/85)
- Schiffrin, D. (1987). *Discourse markers*. London: Cambridge University Press.      
- Schiffrin, D. (2001). [Discourse markers: Language, meaning, and context](http://goo.gl/KbwG9I). In D. Schiffrin, D., D. Tannen, & H. E. Hamilton, (Eds.) *The handbook of discourse analysis* (pp. 54-75). Malden, MA: Blackwell Publishing.
- Ward, N. (2000, October). [Issues in the Transcription of English Conversational Grunts](http://www.aclweb.org/anthology/W00-1004). *Proceedings of the 1st SIGdial Workshop on Discourse and Dialogue* (pp. 29&mdash;35). Hong Kong: Association for Computational Linguistics.  Retrieved from [http://www.aclweb.org/anthology/W00-1004](http://www.aclweb.org/anthology/W00-1004)


## Installation

To download the development version of discon:

Download the [zip ball](https://github.com/trinker/discon/zipball/master) or [tar ball](https://github.com/trinker/discon/tarball/master), decompress and run `R CMD INSTALL` on it, or use the **pacman** package to install the development version:

```r
if (!require("pacman")) install.packages("pacman")
pacman::p_load_gh(
    "trinker/qdapRegex", 
    "trinker/qdapTools", 
    "trinker/qdapDictionaries", 
    "trinker/qdap",
    "trinker/discon"
)
```

## Help

- [Web Page](http://trinker.github.com/discon/)     
- [Package PDF Help Manual](https://dl.dropboxusercontent.com/u/61803503/discon.pdf)   

## Contact

You are welcome to:
* submit suggestions and bug-reports at: <https://github.com/trinker/discon/issues>
* send a pull request on: <https://github.com/trinker/discon/>
* compose a friendly e-mail to: <tyler.rinker@gmail.com>

## List of Discourse Connector Functions

1. `discourse_connector` & `discourse_connector_logical`
2. `dc_backchannel`
3. `dc_causality` & `dc_causality_sub`
4. `dc_comparison`
5. `dc_connective` & `dc_connective_sub`
7. `dc_context` & `dc_context_sub`
7. `dc_equality` & `dc_equality_sub`
8. `dc_filled_pause`
9. `dc_negator`
10. `dc_oh` & `dc_oh_begin`
11. `dc_revision`
12. `dc_timing`
13. `dc_typology`
14. `dc_well` & `dc_well_begin`
15. `kwic` (Key Words in Context)



\****Note*** *that all `discourse_connector` based functions (incuding functions prefixed with `dc_`) have generic `plot` method that utilizes `qdap::dispersion_plot` to generate a lexical dispersion plot.*

## Examples



### Specific Discourse Connector/Connector Functions

Specific discourse connector functions for exploring the context in which such markers are used.  Note these functions are prefixed with a `dc_` (for discourse connector).

<h4 id = "causality" >Causality</h4>



```r
causality_1 <- with(pres_debates2012, dc_causality(dialogue, person))
plot(causality_1)
```

![plot of chunk unnamed-chunk-3](inst/figure/unnamed-chunk-3-1.png) 


```r
plot(causality_1[[1]], high = "darkgreen")
```

![plot of chunk unnamed-chunk-4](inst/figure/unnamed-chunk-4-1.png) 

```r
causality_1[1]
```

```
$counts
     person word.count  causality
1     OBAMA      18317 220(1.20%)
2    ROMNEY      19923  148(.74%)
3   CROWLEY       1670  24(1.44%)
4    LEHRER        765    4(.52%)
5  QUESTION        583    3(.51%)
6 SCHIEFFER       1445   13(.90%)
```

```r
head(causality_1[[2]])
```

```
===================================
Event 1: [lines 5-7]

    LEHRER: And what about the vouchers?

 ** ROMNEY: <<So that>>'s that's number one.

    ROMNEY: Number two is for people coming along that are young,
            what I do to make sure that we can keep Medicare in place
            for them is to allow them either to choose the current
            Medicare program or a private plan. 
 
===================================
Event 2: [lines 9-11]

    ROMNEY: They get to choose and they'll have at least two plans
            that will be entirely at no cost to them.

 ** ROMNEY: <<So>> they don't have to pay additional money, no
            additional dollar six thousand.

    ROMNEY: That's not going to happen. 
```

<h4 id = "causality2" >Causality: Subgroups</h4>


```r
causality_2 <- with(pres_debates2012, dc_causality_sub(dialogue, person))
plot(causality_2, bg.color = "grey60", color = "blue", 
    total.color = "grey40", horiz.color="grey20")
```

![plot of chunk unnamed-chunk-5](inst/figure/unnamed-chunk-5-1.png) 

```r
causality_2[1]
```

```
$counts
     person word.count continuation elaboration
1     OBAMA      18317    127(.69%)    93(.51%)
2    ROMNEY      19923     93(.47%)    55(.28%)
3   CROWLEY       1670     11(.66%)    13(.78%)
4    LEHRER        765      4(.52%)           0
5  QUESTION        583      1(.17%)     2(.34%)
6 SCHIEFFER       1445      7(.48%)     6(.42%)
```

```r
head(causality_2[[2]])
```

```
===================================
Event 1: [lines 5-7]

    LEHRER: And what about the vouchers?

 ** ROMNEY: <<So that>>'s that's number one.

    ROMNEY: Number two is for people coming along that are young,
            what I do to make sure that we can keep Medicare in place
            for them is to allow them either to choose the current
            Medicare program or a private plan. 
 
===================================
Event 2: [lines 9-11]

    ROMNEY: They get to choose and they'll have at least two plans
            that will be entirely at no cost to them.

 ** ROMNEY: <<So>> they don't have to pay additional money, no
            additional dollar six thousand.

    ROMNEY: That's not going to happen. 
```

```r
head(causality_2[[3]])
```

```
===================================
Event 1: [lines 106-108]

    OBAMA: Now, it wasn't just on Wall Street.

 ** OBAMA: You had loan officers were that were giving loans and
           mortgages that really shouldn't have been given,
           <<because>> the folks didn't qualify.

    OBAMA: You had people who were borrowing money to buy a house
           that they couldn't afford. 
 
===================================
Event 2: [lines 120-122]

    OBAMA: And so the question is: Does anybody out there think that
           the big problem we had is that there was too much
           oversight and regulation of Wall Street?

 ** OBAMA: <<Because>> if you do, then Governor Romney is your
           candidate.

    OBAMA: But that's not what I believe. 
```

<h4 id="revision">Revision</h4>


```r
revision <- with(pres_debates2012[1:20, ], dc_revision(dialogue, person))
revision[1]
```

```
$counts
  person word.count revision
1 ROMNEY        260  2(.77%)
2 LEHRER         24 1(4.17%)
```

```r
revision[2]
```

```
$revision
===================================
Event 1: [lines 1-3]

    LEHRER: We'll talk about specifically about health care in a
            moment.

 ** LEHRER: <<But>> what do you support the voucher system, Governor?

    ROMNEY: What I support is no change for current retirees and near
            retirees to Medicare. 
 
===================================
Event 2: [lines 17-19]

    ROMNEY: If I don't like them, I can get rid of them and find a
            different insurance company.

 ** ROMNEY: <<But>> people make their own choice.

    ROMNEY: The other thing we have to do to save Medicare? 
 
===================================
Event 3: [lines 19-20]

    ROMNEY: The other thing we have to do to save Medicare?

 ** ROMNEY: We have to have the benefits high for those that are low
            income, <<but>> for higher income people, we're going to
            have to lower some of the benefits. 
```

<h4 id="typology">Typology</h4>


```r
typology <- with(pres_debates2012[1:120, ], dc_typology(dialogue, person))
typology[1]
```

```
$counts
  person word.count positive negative
1  OBAMA        510  1(.20%)        0
2 ROMNEY        718  3(.42%)        0
3 LEHRER        203  1(.49%)  1(.49%)
```

```r
typology[2]
```

```
$positive
===================================
Event 1: [lines 41-43]

    OBAMA:  And so|

 ** ROMNEY: That's <<that's a>> big topic.

    ROMNEY: Can we can we stay on Medicare? 
 
===================================
Event 2: [lines 98-100]

    ROMNEY: You need transparency, you need to have leverage limits
            for|

 ** LEHRER: Well, <<here's a>> specific|

    ROMNEY: But let's let's mention let me mention the other one. 
 
===================================
Event 3: [lines 103-105]

    LEHRER: Let's let him respond let's let him respond to this
            specific on Dodd Frank and what the governor just said.

 ** OBAMA:  I think this <<is a>> great example.

    OBAMA:  The reason we have been in such a enormous economic
            crisis was prompted by reckless behavior across the
            board. 
```

```r
typology[3]
```

```
$negative
===================================
Event 1: [lines 66-68]

    LEHRER: Beginning with you.

 ** LEHRER: This <<is not a>> new two minute segment to start.

    LEHRER: And we'll go for a few minutes, and then we're going to
            go to health care, OK? 
```

### Generalizable Discourse Connector Function

We will now examine `discourse_connector`, the basic root function that all of the other `dc_` prefixed functions are based upon.  This can be used by the researcher to find specific *text elements* + *surrounding context*.


```r
## Marker with one type (just: "I")
i_disc <- with(pres_debates2012, discourse_connector(dialogue, person,
    names = c("I"),
    regex = "\\bI('[a-z]+)*\\b",
    terms = list(I = c(" I ", " I'"))
))
i_disc[[1]]
```

```
     person word.count          I
1     OBAMA      18317 330(1.80%)
2    ROMNEY      19923 505(2.53%)
3   CROWLEY       1670  49(2.93%)
4    LEHRER        765   9(1.18%)
5  QUESTION        583  18(3.09%)
6 SCHIEFFER       1445  25(1.73%)
```

```r
head(i_disc[[2]])
```

```
===================================
Event 1: [lines 2-4]

    LEHRER: But what do you support the voucher system, Governor?

 ** ROMNEY: What <<I>> support is no change for current retirees and
            near retirees to Medicare.

    ROMNEY: And the president supports taking dollar seven hundred
            sixteen billion out of that program. 
 
===================================
Event 2: [lines 6-8]

    ROMNEY: So that's that's number one.

 ** ROMNEY: Number two is for people coming along that are young,
            what <<I>> do to make sure that we can keep Medicare in
            place for them is to allow them either to choose the
            current Medicare program or a private plan.

    ROMNEY: Their choice. 
```

```r
## Marker with two types (both: "I" & "you")
i_you_disc <- with(pres_debates2012, discourse_connector(dialogue, person,
    names = c("I", "you"),
    regex =  list(
        I = "I('[a-z]+)*\\b",
        you = "(\\b[Yy]ou('[a-z]+)*\\b)"
    ),
    terms = list(
        I = c(" I ", " I'"),
        you = c(" you ", " you'")
    )
))
i_you_disc[[1]]
```

```
     person word.count          I        you
1     OBAMA      18317 330(1.80%) 252(1.38%)
2    ROMNEY      19923 505(2.53%) 227(1.14%)
3   CROWLEY       1670  49(2.93%)  73(4.37%)
4    LEHRER        765   9(1.18%)  36(4.71%)
5  QUESTION        583  18(3.09%)  22(3.77%)
6 SCHIEFFER       1445  25(1.73%)  61(4.22%)
```

```r
head(i_you_disc[[2]])
```

```
===================================
Event 1: [lines 2-4]

    LEHRER: But what do you support the voucher system, Governor?

 ** ROMNEY: What <<I>> support is no change for current retirees and
            near retirees to Medicare.

    ROMNEY: And the president supports taking dollar seven hundred
            sixteen billion out of that program. 
 
===================================
Event 2: [lines 6-8]

    ROMNEY: So that's that's number one.

 ** ROMNEY: Number two is for people coming along that are young,
            what <<I>> do to make sure that we can keep Medicare in
            place for them is to allow them either to choose the
            current Medicare program or a private plan.

    ROMNEY: Their choice. 
```

```r
head(i_you_disc[[3]])
```

```
===================================
Event 1: [lines 1-3]

    LEHRER: We'll talk about specifically about health care in a
            moment.

 ** LEHRER: But what do <<you>> support the voucher system, Governor?

    ROMNEY: What I support is no change for current retirees and near
            retirees to Medicare. 
 
===================================
Event 2: [lines 29-31]

    OBAMA:  That's what they do.

 ** OBAMA:  And so <<you've>> got higher administrative costs, plus
            profit on top of that.

    OBAMA:  And if you are going to save any money through what
            Governor Romney's proposing, what has to happen is, is
            that the money has to come from somewhere. 
```

### Key Words in Context



```r
keyWords <- with(pres_debates2012, kwic(dialogue, list(time, person)))
plot(keyWords[[1]], high = "red")
```

![plot of chunk unnamed-chunk-9](inst/figure/unnamed-chunk-9-1.png) 


```r
with(pres_debates2012, plot(keyWords, grouping.var = person, rm.vars = time, 
    total.color = NULL, bg.color = "black", color = "yellow", 
    horiz.color="grey20"))
```

![plot of chunk unnamed-chunk-10](inst/figure/unnamed-chunk-10-1.png) 

```r
head(keyWords[[2]])
```

```
===================================
Event 1: [lines 3-5]

    time 1.ROMNEY: What I support is no change for current retirees
                   and near retirees to Medicare.

 ** time 1.ROMNEY: And the <<president>> supports taking <<dollar>>
                   seven hundred sixteen billion out of that program.

    time 1.LEHRER: And what about the vouchers? 
 
===================================
Event 2: [lines 9-11]

    time 1.ROMNEY: They get to choose and they'll have at least two
                   plans that will be entirely at no cost to them.

 ** time 1.ROMNEY: So they don't have to pay additional money, no
                   additional <<dollar>> six <<thousand>>.

    time 1.ROMNEY: That's not going to happen. 
```





NEWS
NEWS 
====

Versioning
----------

Releases will be numbered with the following semantic versioning format:

<major>.<minor>.<patch>

And constructed with the following guidelines:

* Breaking backward compatibility bumps the major (and resets the minor 
  and patch)
* New additions without breaking backward compatibility bumps the minor 
  (and resets the patch)
* Bug fixes and misc changes bumps the patch

discon 0.2.0  
----------------------------------------------------------------

BUG FIXES

NEW FEATURES

* `dc_connective` & `dc_connective_sub` functions added to search for conectives
  ("and", "but", & "or") in context.

MINOR FEATURES

* `discourse_connector_logical` now takes a `fun3` and `fun4` argument that 
  allows the user to operate on `text.var` or just `text.var2`.  These are 
  undocumented and typically for internal use.  This move is an attempt to make
  the function more modular, for example `dc_comparison` now takes advantage of 
  this template.

IMPROVEMENTS

* `dc_comparison` now utilizes `discourse_connector_logical` + the internal 
  `hijack` rather than a less maintainable separate function.

CHANGES


discon 0.1.0  
----------------------------------------------------------------

BUG FIXES

* `dicourse_connector` would error if a regex did not match any indices.  It now
  returns `NULL` and gives a message indicating what regex did not match any 
  indices.

NEW FEATURES

* `dc_negator` & `dc_negator_sub` functions added to search for negator discourse 
  connectors.

* `dc_comparison` function added to search for comparison discourse connectors.

* `dc_filled_pause` function added to search for filled pause discourse connectors.

* `dc_oh` function added to search for oh discourse connectors (more specifically 
  discourse marker).  The `dc_oh_begin` function adds the assertion that the "oh"
  usage is at the beginning of a line (this is a sentence if `qdap::sentSplit` has
  been used).

* `dc_well` function added to search for oh discourse connectors (more specifically 
  discourse marker).  The `dc_well_begin` function adds the assertion that the "well"
  usage is at the beginning of a line (this is a sentence if `dap::sentSplit` has
  been used).

* `dc_backchannel` function added to search for backchannel discourse connectors. 

* `kwic` functions added to search for  and plot the distribution of 
  *key words in context*.

* `discourse_connector_logical` added that allows the extension of 
  `discourse_connector` to enable logical checks of the elements of the 
  `text.var` and/or the `grouping.var`.  The ease of use for `qdap::termco` and 
  `qdap::dispersion_plot` makes them less flexible than a pure regular expression 
  approach.  `discourse_connector_logical` returns much of this flexibility
  back to the `discourse_connector` base function.  An example use would be to
  assert that the word "oh"/"Oh" occurs at the beginning of a line or a turn of
  talk.  The user can pass functions to assert this.

* Logical testing functions added to complement the `discourse_connector_logical`
  function.  `is.first_in_set` (and `is.last_in_set` as the semantical inverse), 
  `is.isolate` (and `is.repeat` as the semantical inverse), and 
  `is.within_n_preceding_words` added to assert that elements in discourse or 
  group variables are patterned win a particular way.  These functions are useful
  as a `discourse_connector_logical` complement or as stand alone discourse 
  analysis tools.

IMPROVEMENTS

* Generic `plot` for `dicourse_connector` picks up a `grouping.var` argument.
  This allows for flexibility in use while maintaining the no fuss default 
  settings.

CHANGES

* The **discmark** package was renamed to **discon**.  The latter is a more 
  inclusive term that better defines the scope of the package.


discon 0.0.3   
----------------------------------------------------------------

BUG FIXES

* `discourse_connector` did not retain the correct grouping variable(s) name in the
  `termco` output as well as the `ggplot::ylab` on the `dispersion_plot`.  

NEW FEATURES

* `dc_context` & `dc_context_sub` functions added to search for context discourse 
  connectors.

* `dc_equality` & `dc_equality_sub` functions added to search for equality 
  discourse connectors.

* `dc_timing` function added to search for timing discourse connectors.

IMPROVEMENTS

* `additional unit test for `print.discourse_connector` added.

CHANGES

* **ggplot2** added to the DESCRIPTION file's `Imports:` for correction of the 
  plot's `ylab` in the plot method for `discourse_connector`.



discon 0.0.2   
----------------------------------------------------------------

NEW FEATURES

* `discourse_connector` template added to make the package more modular and 
  consistent.  `discoure_connector` also becomes a generic class transferred to all 
  `dc_` prefixed discourse connector functions.  The accompanying plot method 
  allows for a lexical dispersion plot, a visual method of seeing distribution of 
  terms across time.

* `dc_causality` & `dc_causality_sub` functions added to search for causal discourse 
  connectors.

* `dc_revision` function added to search for revision discourse connectors.

IMPROVEMENTS

* `hijack` function used with `discourse_connector` template to make the package 
  more modular.

* `discourse_connector` is expanded to generalize to n lists of regular expressions 
  (`regex`) rather than 1-2.  The `regex.markup` argument has been removed and
  `regex` supplies this information.

* Examples added to the **README.md** fie.

CHANGES

* `taxonomy` renamed to `dc_typology`.


discon 0.0.1 
----------------------------------------------------------------

Initial upload of **discon**:

**discon** is a collection of tools for analysing discourse connectors in text. 



R/kwic.R
#' Extract Key Words in Context
#'
#' Extract key words in context (kwic) utilizing \code{\link[qdap]{freq_terms}}.
#' 
#' @param text.var The text variable. 
#' @param grouping.var The grouping variables.  Also takes a single 
#' grouping variable or a list of 1 or more grouping variables.
#' @param n.before The number of rows before the indexed occurrence.
#' @param tot logical.  If \code{TRUE} condenses sub-units (e.g., sentences) 
#' into turns of talk for that \code{grouping.var}.
#' @param n.after The number of rows after the indexed occurrence.
#' @param ord.inds logical.  If \code{TRUE} inds is ordered least to greatest.
#' @param markup A character vector of length two indicating the left (element 
#' 1) and right (element 2) boundary markers to use to highlight the 
#' key words.  Use \code{c("", "")} to not mark the discourse 
#' connectors.
#' @param top Top number of terms to show.
#' @param at.least An integer indicating at least how many letters a word 
#' must be to be included in the output.
#' @param extend logical.  If \code{TRUE} the \code{top} argument is extended to 
#' any word that has the same frequency as the \code{top} word.
#' @param ignore.case logical.  If \code{TRUE} leading character's case is 
#' ignored. 
#' @param left The left boundary (regular expression) to put on the left side of 
#' words.  Default is 
#' \href{http://www.regular-expressions.info/wordboundaries.html}{word boundary} 
#' (i.e., \code{\\\\b}) that cuts off at non-word characters.  
#' @param right The right boundary (regular expression) to put on the right side 
#' of words.  Default is no boundary and will locate words within words.  For 
#' example the expression \code{"\\\\blast"} will find "last" and "lastly" but not 
#' "blast".  To capture precisely key words use \code{right = "\\\\b"} or 
#' \code{right = left}.
#' @param names A string indicating the name of the key words in the collective
#' group name for the key words for use in the \code{\link[qdap]{termco}} and 
#' generic \code{plot} (\code{\link[qdap]{dispersion_plot}}) functions.
#' @param elim.old logical.  If \code{TRUE} eliminates the columns that are 
#' combined together by the named match.list.
#' @param stopwords A character vector of words to remove from the text.
#' @param \ldots Other arguments passed to \code{\link[qdap]{termco}}.
#' @details While \code{kwic} does not analyze discourse connectors, it is useful
#' in identifying themes (and distribution across time) and builds upon the 
#' modular use of functions in the pkg{discon} package.
#' @return Returns returns a list of 2-3: 
#' \item{counts}{A \code{\link[qdap]{termco}} object of discourse connector counts.} 
#' \item{Context 1}{A \code{\link[qdap]{trans_context}} object of the key words in context.  Note the name of this object is supplied by \code{names} element one.} 
#' \item{Context 2...n}{An optional (not returned if \code{regex} is of length one) \code{\link[qdap]{trans_context}} object of the key words in context.  Note the name of this (these) object(s) is supplied by \code{names} element 2...n.} 
#' @references Ryan, G. W. & Bernard, H. R. (2003). \href{http://goo.gl/KdYxB8}{Techniques to identify themes}. \emph{Field Methods. 15}(1), 85-109. doi: \href{http://fmx.sagepub.com/content/15/1/85}{10.1177/1525822X02239569}
#'
#' @keywords key words frequent kwic
#' @export
#' @importFrom qdap trans_context termco
#' @seealso \code{\link[qdap]{termco}},
#' \code{\link[qdap]{trans_context}},
#' \code{\link[qdap]{freq_terms}}
#' @rdname kwic
#' @include utils.R internal_data_bases_list.R internal_data_regex_list.R internal_data_term_list.R
#' @examples
#' out <- with(pres_debates2012, kwic(dialogue, list(time, person)))
#' plot(out[[1]])
#' head(out[[2]])
#' with(pres_debates2012, plot(out, grouping.var = person, rm.vars = time, 
#'     total.color = NULL))
#' 
#' ## Save externally use .doc or .txt
#' ## print(out[[2]], file="kwic.doc")
kwic <- function(text.var, grouping.var, n.before = 1, tot = FALSE, 
    n.after = n.before, ord.inds = TRUE, markup = c("<<", ">>"), top = 15,
    at.least = 5, extend = TRUE, ignore.case = FALSE, left = "\\b", right = "",
    names = c("KeyWords"), elim.old = FALSE, 
    stopwords = c("going", qdapDictionaries::contractions[[1]], 
        qdapDictionaries::Top200Words), ...){

    ## Grab the grouping variable name
    if (is.list(grouping.var)) {
        m <- unlist(as.character(substitute(grouping.var))[-1])
        m <- sapply(strsplit(m, "$", fixed=TRUE), function(x) {
                x[length(x)]
            }
        )
        group.nms <- paste(m, collapse="&")
    } else {
        G <- as.character(substitute(grouping.var))
        group.nms <- G[length(G)]
    }

    ## Determine top n words
    topterms <- qdap::freq_terms(
        text.var, 
        top = top, 
        at.least = at.least, 
        stopwords = stopwords, extend = extend
    )
    
    ## Marker with top 15 words
    out <- discourse_connector(text.var, grouping.var,
        names = names,
        regex =  setNames(list(
            binder(topterms[[1]], ignore.case = ignore.case, 
                left = left, right = right)
        ), names),
        terms = setNames(list(
            qdap::spaste(topterms[[1]])
        ), names), elim.old = elim.old,
    )
    out[["counts"]] <- termco_group_name_replace(out[["counts"]], group.nms)
    attributes(out)[["meta"]][["group.nms"]] <- group.nms
    out
}

